middle_end/flambda2/simplify/simplify_expr.ml

```diff
@@ -51,23 +51,22 @@ and simplify_toplevel dacc expr ~return_continuation ~return_arity
     DA.map_data_flow dacc ~f:(Data_flow.init_toplevel dummy_toplevel_cont [])
   in
   let expr, uacc =
-    simplify_expr dacc expr ~down_to_up:(fun dacc ~rebuild ->
+    simplify dacc ~down_to_up:(fun dacc ~rebuild ->
         let dacc =
           DA.map_data_flow dacc
             ~f:(Data_flow.exit_continuation dummy_toplevel_cont)
         in
         let data_flow = DA.data_flow dacc in
-        let closure_info = DE.closure_info (DA.denv dacc) in
         (* The code_age_relation and used value_slots are only correct at
            toplevel, and they are only necessary to compute the live code ids,
            which are only used when simplifying at the toplevel. So if we are in
            a closure, we use empty/dummy values for the code_age_relation and
            used_value_slots, and in return we do not use the reachable_code_id
            part of the data_flow analysis. *)
         let code_age_relation, used_value_slots =
-          match Closure_info.in_or_out_of_closure closure_info with
-          | In_a_closure -> Code_age_relation.empty, Or_unknown.Unknown
-          | Not_in_a_closure ->
+          if is_in_closure
+          then Code_age_relation.empty, Or_unknown.Unknown
+          else
```

Comment:
I tried to do this, but the type for this is private. I'm not sure what would be best, either leave a boolean, make the type not private, or export the constructors (in which case the obvious names are already taken).

Reply:
I did this change: https://github.com/ocaml-flambda/flambda-backend/pull/836/commits/9d8d358eea2fe6d3aa0d04e9a9b6cf45e471604e
However, I'm not convinced about it, since we need to specify the type everywhere it is used to avoid a non-principal type disambiguation, so if you have a better option, I'm all for it.